### PR TITLE
Improve speed to run `airflow` by 6x

### PR DESCRIPTION
**Pull request by ashb**
By delaying expensive/slow imports to where they are needed, this gets `airflow` printing it's usage information in under 0.5s, down from almost 3s which makes it feel much much snappier.

I can't really think of any way of adding tests to prevent regressions here. Ideas welcome.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

File: airflow/exceptions.py [hunk 1]
```diff
@@ -44,7 +40,7 @@ class AirflowBadRequest(AirflowException):
     status_code = 400
 
 
-class AirflowNotFoundException(AirflowException, ApiConnexionNotFound):
```
Comment (ashb) [inline airflow/exceptions.py:47]: I have a feeling that this might break something. Lets see.

**Comment by ashb** [comment]
This was profiled using `python -X importtime -m airflow` and then looking for high cumulative modules times (3rd column)

File: airflow/executors/executor_loader.py [hunk 2]
```diff
@@ -35,6 +34,9 @@
 
 log = logging.getLogger(__name__)
 
+if TYPE_CHECKING:
+    from airflow.executors.base_executor import BaseExecutor
```
Comment (ashb) [inline airflow/executors/executor_loader.py:38]: By moving this here it prevents loading all the models and saves a goodly chunk of time.

**Comment by potiuk** [comment]
I love the "perceived speed" factor. You can't imagine how many tricks mobile apps do to load "percivably faster". And to be perfectly hones this is one of the most important factors that improve "quality" of the apps.

File: airflow/exceptions.py (hunk 1 above)
Comment (potiuk) [reply airflow/exceptions.py:47]: Same here. Let's see :)

File: airflow/executors/executor_loader.py (hunk 2 above)
Comment (potiuk) [reply airflow/executors/executor_loader.py:38]: Fantastic!

**Comment by ashb** [comment]
(This was totally not what I "meant" to be work on tonight. But sometimes you have to roll with it)

**Review by potiuk — APPROVED** [review]
Let's see - but it looks good.

BTW. I have a vague feeling this is really a band-aid and we need to refactor the modules to avoid local imports altogether. For me the need to add "TYPE_CHECKING" and local imports is a sign we have the structure of "airflow" package wrong. 

However it is likely unsolvable until airfliow as it might lead to breaking compatibility, so for now it is "good enoug".

**Comment by ashb** [comment]
> Let's see - but it looks good.
> 
> BTW. I have a vague feeling this is really a band-aid and we need to refactor the modules to avoid local imports altogether. For me the need to add "TYPE_CHECKING" and local imports is a sign we have the structure of "airflow" package wrong.
> 
> However it is likely unsolvable until airfliow as it might lead to breaking compatibility, so for now it is "good enoug".

Yeah no disagreement from me -- and primarirly one of the thngs I have always not liked about the code is the number of side-effects we get from just `import airflow`

**Comment by potiuk** [comment]
@Ashb - seems like you've been ahead of our users reporting this: https://github.com/apache/airflow/discussions/21802#discussioncomment-2452256 (connected to the fact that someone's filesystem is SLOW)

**Pull request merged**